### PR TITLE
[MPS] Add SDPA implentation

### PR DESCRIPTION
**Pull request by manuelcandales**
This work is based off @malfet's #119200

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -14724,6 +14724,11 @@
   variants: function
   tags: nondeterministic_seeded
 
+- func: _scaled_dot_product_attention_math_for_mps(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool is_causal=False, Tensor? dropout_mask=None, *, float? scale=None) -> (Tensor, Tensor)
```
Comment (kimishpatel) [inline aten/src/ATen/native/native_functions.yaml:14727]: Why add to native_functions.yaml

Comment (manuelcandales) [reply aten/src/ATen/native/native_functions.yaml:14727]: So that I can compare against _scaled_dot_product_attention_math on Metal

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    MPSGraphTensor* vTensor = nil;
+    MPSGraphTensor* attnTensor = nil;
+    MPSGraphTensor* outputTensor = nil;
+  };
+  int64_t batchSize = query.size(0);
+  int64_t qSize = query.size(2);
+  int64_t num_head = query.size(1);
+  int64_t headSize = query.size(3);
+  auto out = at::empty({batchSize, num_head, qSize, headSize}, query.options());
+  auto scale_factor = sdp::calculate_scale(query, scale).sqrt().as_float_unchecked();
```
Comment (DenisVieriu97) [inline aten/src/ATen/native/mps/operations/Attention.mm:43]: fyi, there's a native op in MPSGraph starting with Sequoia https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/scaleddotproductattentionwithquerytensor:keytensor:valuetensor:scale:name:  (masked one: https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/scaleddotproductattentionwithquerytensor:keytensor:valuetensor:masktensor:scale:name:)

Comment (kimishpatel) [reply aten/src/ATen/native/mps/operations/Attention.mm:43]: says the page cannot be found?

BTW is the native op implementation of flash attention?

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    auto mkey = __func__ + getTensorsStringKey({query, key, value}) + ":" + std::to_string(is_causal) + ":" + std::to_string(attn_mask.has_value());
+    auto cachedGraph = LookUpOrCreateCachedGraph<CachedGraph>(mkey, [&](auto mpsGraph, auto graph) {
+      auto qTensor = mpsGraphRankedPlaceHolder(mpsGraph, query);
+      auto kTensor = mpsGraphRankedPlaceHolder(mpsGraph, key);
+      auto vTensor = mpsGraphRankedPlaceHolder(mpsGraph, value);
+      auto kT = [mpsGraph transposeTensor:kTensor dimension:2 withDimension:3 name:nil];
+      auto scaleTensor = [mpsGraph constantWithScalar:scale_factor shape:getMPSShape({1}) dataType:MPSDataTypeFloat32];
+
+      // sqTensor = qTensor * scale_factor
+      auto castedQ = qTensor;
```
Comment (kimishpatel) [inline aten/src/ATen/native/mps/operations/Attention.mm:57]: Leave a comment on why this upcasting

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      if ([qTensor dataType] != MPSDataTypeFloat32) {
```
Comment (kimishpatel) [inline aten/src/ATen/native/mps/operations/Attention.mm:58]: Also do you wanna assert on supported dtypes being fp16 and fp32 only? or bf16 is also supported?

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      if ([sqTensor dataType] != qTensor.dataType) {
+        sqTensor = [mpsGraph castTensor:sqTensor toType:qTensor.dataType name:nil];
+      }
+
+      // skT = kT * scale_factor
+      auto castedKT = kT;
+      if ([kT dataType] != MPSDataTypeFloat32) {
+        castedKT = [mpsGraph castTensor:kT toType:MPSDataTypeFloat32 name:nil];
+      }
+      auto skT = [mpsGraph multiplicationWithPrimaryTensor:castedKT secondaryTensor:scaleTensor name:nil];
```
Comment (kimishpatel) [inline aten/src/ATen/native/mps/operations/Attention.mm:71]: Why are you not doing `q @ k.T` and then multiply by scale

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+      if ([skT dataType] != kT.dataType) {
+        skT = [mpsGraph castTensor:skT toType:kT.dataType name:nil];
+      }
+
+      auto maskedMM = [mpsGraph matrixMultiplicationWithPrimaryTensor:sqTensor secondaryTensor:skT name:nil];
+      if (is_causal) {
```
Comment (kimishpatel) [inline aten/src/ATen/native/mps/operations/Attention.mm:77]: assert that it cannot support both is_causal and attn_mask together

File: aten/src/ATen/native/transformers/attention.cpp [hunk 7]
```diff
@@ -985,5 +996,5 @@ Tensor triton_multi_head_attention(
 #endif
   return proj;
 }
-} // namespace native
-} // namespace at
```
Comment (kimishpatel) [inline aten/src/ATen/native/transformers/attention.cpp:989]: why this change?

File: test/test_mps.py [hunk 8]
```diff
@@ -9227,8 +9227,73 @@ def weight_int8pack_mm(a, b_int8pack, b_scales):
...
+    def _test_sdpa_no_mask(self, is_causal: bool):
+        torch.manual_seed(1729)
+        max_seq_length = 72
+        with torch.nn.attention.sdpa_kernel([torch.nn.attention.SDPBackend.MATH]):
+            q = torch.randn([1, 32, 1, 128], dtype=torch.float16, device="mps")
+            k = torch.randn([1, 32, max_seq_length, 128], dtype=q.dtype, device="mps")
+            v = torch.randn([1, 32, max_seq_length, 128], dtype=q.dtype, device="mps")
+                       
+            y, attn = torch.ops.aten._scaled_dot_product_attention_math_for_mps(q, k, v, dropout_p=0.0, is_causal=is_causal)
+            y_ref, attn_ref = torch.ops.aten._scaled_dot_product_attention_math(q, k, v, dropout_p=0.0, is_causal=is_causal)
```
Comment (kimishpatel) [inline test/test_mps.py:9245]: Now you can compare against cpu impl as well

File: test/test_mps.py [hunk 9]
```diff
@@ -9227,8 +9227,73 @@ def weight_int8pack_mm(a, b_int8pack, b_scales):
...
+        max_seq_length = 72
+        causal_mask = torch.tril(
+            torch.ones(max_seq_length, max_seq_length, dtype=torch.bool, device='mps'),
+        )
+        with torch.nn.attention.sdpa_kernel([torch.nn.attention.SDPBackend.MATH]):
+            i = 42
+
+            q = torch.randn([1, 32, 1, 128], dtype=torch.float16, device="mps")
+            k = torch.randn([1, 32, max_seq_length, 128], dtype=q.dtype, device="mps")
+            v = torch.randn([1, 32, max_seq_length, 128], dtype=q.dtype, device="mps")
```
Comment (kimishpatel) [inline test/test_mps.py:9267]: Some prime sizes on the tensor?

Comment (manuelcandales) [reply test/test_mps.py]: done!

File: test/test_mps.py (hunk 8 above)
Comment (manuelcandales) [reply test/test_mps.py]: done!

File: aten/src/ATen/native/transformers/attention.cpp (hunk 7 above)
Comment (manuelcandales) [reply aten/src/ATen/native/transformers/attention.cpp:989]: done!

File: aten/src/ATen/native/mps/operations/Attention.mm (hunk 6 above)
Comment (manuelcandales) [reply aten/src/ATen/native/mps/operations/Attention.mm:74]: done!

File: aten/src/ATen/native/mps/operations/Attention.mm (hunk 5 above)
Comment (manuelcandales) [reply aten/src/ATen/native/mps/operations/Attention.mm]: done!

File: aten/src/ATen/native/mps/operations/Attention.mm (hunk 4 above)
Comment (manuelcandales) [reply aten/src/ATen/native/mps/operations/Attention.mm]: bf16 is also supported

File: aten/src/ATen/native/mps/operations/Attention.mm (hunk 3 above)
Comment (manuelcandales) [reply aten/src/ATen/native/mps/operations/Attention.mm]: done!

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+                                                                  const Tensor& value,
+                                                                  const std::optional<Tensor>& attn_mask,
+                                                                  double dropout_p,
+                                                                  bool is_causal,
+                                                                  const std::optional<Tensor>& dropout_mask,
+                                                                  std::optional<double> scale) {
+  if (is_causal) {
+    TORCH_CHECK(!attn_mask.has_value(),
+            "_scaled_dot_product_attention: Explicit attn_mask should not be set when is_causal=True");
+    TORCH_CHECK(!query.is_nested() && !key.is_nested(),
```
Comment (kimishpatel) [inline aten/src/ATen/native/mps/operations/Attention.mm:33]: Is it supported when is_causal=False?

File: test/test_mps.py [hunk 11]
```diff
@@ -9250,8 +9250,68 @@ def weight_int8pack_mm(a, b_int8pack, b_scales):
...
+class TestSDPA(TestCaseMPS):
+    def _compare_tensors(self, y, ref):
+        denom = torch.maximum(ref.abs(), torch.tensor([1e-6], device=ref.device, dtype=ref.dtype))
+        err = ((y - ref).abs() / denom).mean().item()
+        self.assertLess(err, 0.01)
 
+    def _test_sdpa_no_mask(self, is_causal: bool, dtype: torch.dtype, L: int = 1, S: int = 72):
+        torch.manual_seed(1729)
+        with torch.nn.attention.sdpa_kernel([torch.nn.attention.SDPBackend.MATH]):
+            q = torch.randn([1, 32, L, 128], dtype=dtype, device="mps")
```
Comment (kimishpatel) [inline test/test_mps.py:9262]: also test with different sizes like 23, 121. Not just power of 2.

**Review by kimishpatel — APPROVED** [review]
Left a minor comment

**Review by kimishpatel — APPROVED** [review]
Left a minor comment

File: test/test_mps.py (hunk 11 above)
Comment (manuelcandales) [reply test/test_mps.py]: done!

File: aten/src/ATen/native/mps/operations/Attention.mm (hunk 10 above)
Comment (manuelcandales) [reply aten/src/ATen/native/mps/operations/Attention.mm]: moved out of if clause
